Issue 4956 - Implement search filter for country and state when editing an order

RenanLukas opened the pull request:
<!-- Remember about a good descriptive title. -->

# Implement search for country/state on address editing screen

Closes: #4956
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
Implements a search feature for country and state selection on the address editing screen. Currently we have a Dialog for this selection, and user has to scroll to find the desired option. With the search filtering the items it is possible to find an option faster than scrolling.

It seems that we were allowing user to tap on the state selection even if country was empty. I've disabled the option to select state until country is fulfilled, but if this was on purpose I'll revert the commit (e3cbcbd6fca61bda11e9f1ee2eccbb3ace117bae).

<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1 - Login with an user that has at least one order of a physical product in any status (the order must contain a physical product in order to test all the scenarios);
2 - Navigate to "Orders" screen and tap on any order to open details;

**To test billing address**
3 - Tap on "Add Billing Address" or the edit button if the address is already filled.
4 - Tap on "Country"  to open search filter for country. Select a country;
5 - Tap on "State" to open search filter for state. Select a state;
6 - Tap on "Done" to update the address with the selected options.

**To test shipping address**
3 - Tap on "Add Shipping Address" or the edit button if the address is already filled.
4 - Tap on "Country"  to open search filter for country. Select a country;
5 - Tap on "State" to open search filter for state. Select a state;
6 - Tap on "Done" to update the address with the selected options.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

#### State disabled
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660246-e1a5b2d4-a6ee-4bbd-b734-ed45c7ae6a36.png" alt="State disabled" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661178-cbfecca0-c40f-484c-b102-3bffcc6b54c3.png" alt="State disabled (dark)" width="250" /> 
</p>

#### Searching for country
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660250-0d1574a5-a414-491b-878f-3a07274f635d.png" alt="Searching for country" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661235-66100ea3-10a5-49f5-b9a7-10665373f593.png" alt="Searching for country" width="250" />
</p>

#### Search country with 3 results
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660257-7c4fef1d-ca3a-4dbd-9c98-b18acfce0aed.png" alt="Search country with 3 results" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661251-3b066034-d320-4c56-9e1d-dcc6d23cc37d.png" alt="Search country with 3 results (dark)" width="250" />
</p>

#### Search country with 1 result
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660627-cf894448-fd77-4604-aacf-7c7f48fe1a8c.png" alt="Search country with 1 result" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661266-ba389875-49fb-44fa-8990-3ba4c6c524ce.png" alt="Search country with 1 result (dark)" width="250"/>
</p>

#### Search empty state
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660278-0a4b8187-7597-45e1-ad63-d741208a7e91.png" alt="Search empty state" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661274-4cd3441d-0bc3-4779-b35f-0d8ffbf3b679.png" alt="Search empty state (dark)" width="250"/>
</p>

#### Country selected
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660299-5c99668a-3e4e-41ab-8352-6d7553faa5c3.png" alt="Country selected" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661293-e9af0017-941e-4cae-ba0b-c234220fdf15.png" alt="Country selected (dark)" width="250"/>
</p>

#### State enabled
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660309-c4c271be-7019-496b-94d0-5ec42c7443c4.png" alt="State enabled" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661310-bb03a78d-b92b-4a2c-88cf-1ae3e456a5e3.png" alt="State enabled (dark)" width="250"/>
</p>

#### Search state
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660316-00d1eca6-bd90-4341-95bb-b83f5bce38d0.png" alt="Search state" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661389-80f47874-ba79-481d-9a0f-63a8148a9642.png" alt="Search state (dark)" width="250"/>
</p>

#### Search state with 1 result
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660656-d1b6d844-0496-4918-8b01-04c17752480f.png" alt="Search state with 1 result" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661402-e4959219-04c9-4e92-8598-b653211e0c8f.png" alt="Search state with 1 result (dark)" width="250"/>
</p>

#### Country and state selected
<p float="center">
<img src="https://user-images.githubusercontent.com/14964993/140660330-d600925e-e06c-4861-ba1b-8a56e6a51e69.png" alt="Country and state selected" width="250"/>
<img src="https://user-images.githubusercontent.com/14964993/140661424-fd12d148-7fc8-4afd-8c19-58a65375a2f3.png" alt="Country and state selected (dark)" width="250"/>
</p>


- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

RenanLukas committed:
Create Fragment, layout file and navigation arguments for search filter screen

RenanLukas committed:
Format nav_graph_orders

RenanLukas committed:
Replace current edit country and state action with open search filter screen

RenanLukas committed:
Add search component to search filter screen

RenanLukas committed:
Remove selected item argument from search filter screen

RenanLukas committed:
Extract set title to separate method

RenanLukas committed:
Remove selected item argument from search filter action calls

RenanLukas committed:
Implement items list on search filter screen

RenanLukas committed:
Implement return result on search filter screen

RenanLukas committed:
Implement search on search filter screen

RenanLukas committed:
Add empty view to search filter screen

RenanLukas committed:
Fix search filter background color

RenanLukas committed:
Fix SearchView layout on search filter screen

RenanLukas committed:
Replace SearchView with EditText on search filter screen

RenanLukas committed:
Implement SearchFilterViewModel

RenanLukas committed:
Implement SearchFilterViewModel unit tests

RenanLukas committed:
Add unit tests to AddressViewModel

RenanLukas committed:
Refactor AddressViewModel unit test name

RenanLukas committed:
Implement enable state selection only when country is already selected

RenanLukas committed:
Fix list of items showing on empty state

RenanLukas committed:
Fix country and state view state update on onViewCreated

With the new implementation of a search screen for country/state selection, onViewCreated is triggered when user navigates back to the address editing screen. That was causing the view state country to be replaced with empty string instead of the user selected country. This implementation checks if user selected a country (using addressDraft) and use that to update the view state. If user did not select a country in this instance of address editing screen, addressDraft.country will be empty and we can use storedAddress without problems.

RenanLukas committed:
Fix detekt indentation warnings on SearchFilterFragment

RenanLukas committed:
Fix edit address scroll being reset after navigating back from selecting country/state

nbradbury commented:
Maybe consider using `android:imeOptions="flagNoExtractUi"` on the search view so the screen isn't taken over in landscape?

![land](https://user-images.githubusercontent.com/3903757/140802037-143e99b6-ca06-4167-8742-02c8442de10c.png)

nbradbury commented:
Minor nit: I realize the designs we shared with you came from iOS, but Material Guidelines recommend only capitalizing the first word, so this should be "Filter countries."

![filter](https://user-images.githubusercontent.com/3903757/140803404-2e8628bb-f41a-4356-a37d-b724308349a5.png)

nbradbury commented:
If I go to edit an address in the United States and no state has been selected, if I select one but then back out and discard changes, that state shows up if you go to edit the address again. A similar problem happens if you discard changes after selecting a different country.

I wasn't able to reproduce these in our `develop` so I'm assuming they were introduced in this PR?

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/searchfilter/SearchFilterFragment.kt:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    private fun setupTitle(title: String) {
+        activity?.title = title
```
Just an FYI that you can override `getFragmentTitle()` in `BaseFragment` to supply the activity title, like this:

```
override fun getFragmentTitle() = navArgs.title
```

nbradbury commented on WooCommerce/src/main/res/layout/fragment_search_filter.xml:
```diff
@@ -0,0 +1,54 @@
+            android:layout_marginBottom="@dimen/major_100"
+            android:drawableStart="@drawable/ic_search_24dp"
+            android:drawablePadding="@dimen/major_100"
+            app:layout_constraintEnd_toEndOf="parent"
```
Theses constraints are unnecessary since this is contained in a `FrameLayout`.

RenanLukas replied:
My bad 😅

nbradbury commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/details/editing/address/AddressViewModelTest.kt:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    @Test
+    fun `Should update viewState with selected country, reset selected state and enable state selection on onCountrySelected`() {
```
I'm getting a detekt `MaximumLineLength` error on this.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/searchfilter/SearchFilterAdapter.kt:
```diff
@@ -0,0 +1,47 @@
+
+    fun setItems(items: List<SearchFilterItem>) {
+        this.items = items
+        notifyDataSetChanged()
```
Would it make sense to use `DiffUtil.calculateDiff` here instead of `notifyDataSetChanged`?

RenanLukas replied:
Makes sense 👍

nbradbury commented:
@RenanLukas I've done my first round of reviews. You've done some very nice work here. I like the way you relied on `SearchFilterViewState` and made the filter screen so easy to reuse. 

I left a few comments for you to look into, but apart from the state/country reappearing there's nothing serious to address.

RenanLukas committed:
Add flagNoExtractUi to search EditText to fix layout in landscape

RenanLukas committed:
Fixed search country hint capitalizing the second word

RenanLukas committed:
Simplified search filter title logic

RenanLukas commented:
> Minor nit: I realize the designs we shared with you came from iOS, but Material Guidelines recommend only capitalizing the first word, so this should be "Filter countries."
> 
> ![filter](https://user-images.githubusercontent.com/3903757/140803404-2e8628bb-f41a-4356-a37d-b724308349a5.png)

@nbradbury Thanks for pointing out! 
Should we follow the same approach for the `Filter State / County` string? I've capitalized the words because that's how it looks like in the address edit screen. Would it make sense to change it to `Filter state / county` on the search filter screen and to `State / county` (instead of `State / County`) on the address edit screen?

RenanLukas committed:
Fix discarded state and country appearing in order address edit

RenanLukas commented:
@nbradbury thank you for taking the time to review my PR. I've left a [small question](https://github.com/RenanLukas/woocommerce-android/pull/1#issuecomment-964634986) and have addressed your comments on this first round.

![unit tests](https://user-images.githubusercontent.com/14964993/141048621-b6656682-2682-4ffd-8e2a-1d8b27400eb9.png)

nbradbury commented:
> Should we follow the same approach for the `Filter State / County` string? I've capitalized the words because that's how it looks like in the address edit screen.

@RenanLukas Looking at the iOS app I see it just says "State" (no county) so let's just mimic that in the address editing screen and the filter screens.

nbradbury reviewed (changes requested):
@RenanLukas I'm looking at your PR and noticing that you missed a few comments I made (my mistake).

RenanLukas committed:
Improve SearchFilterAdapter using DiffUtil instead of calling notifyDataSetChanged

RenanLukas committed:
Fix detekt MaximumLineLength on AddressViewModelTest

RenanLukas committed:
Remove unnecessary constraints from searchEditText

RenanLukas committed:
Remove county from edit address state strings

RenanLukas commented:
Hey @nbradbury , thanks for checking it again. I've addressed the other comments you made.

nbradbury reviewed (approved):
Changes look great, very nice work! I'm going to ask Wojtek to give this a quick look and then I think we're good to merge 🎉

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/searchfilter/SearchFilterAdapter.kt:
```diff
@@ -0,0 +1,65 @@
+
+class SearchFilterAdapter(
+    private val onItemSelectedListener: (SearchFilterItem) -> Unit
+) : RecyclerView.Adapter<SearchFilterViewHolder>() {
```
💡  If we use `DiffUtil`, maybe it makes sense to use [`ListAdapter`](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter) class? It computes diff on background thread and make codebase consistent. We already use it in e.g. `ProductFilterOptionListAdapter`. WDYT?

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/searchfilter/SearchFilterViewState.kt:
```diff
@@ -0,0 +1,17 @@
+package com.woocommerce.android.ui.searchfilter
+
+sealed class SearchFilterViewState {
```
💯 nice! I like this interpretation.

wzieba commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/details/editing/address/AddressViewModelTest.kt:
```diff
@@ -0,0 +1,205 @@
+                countryCode = "countryCode",
+                stateCode = "stateCode"
+            )
+            verify(dataStore, times(1)).fetchCountriesAndStates(selectedSite.get())
```
⛏️  minor nitpick: `verify` by default checks for exactly 1 execution
```suggestion
            verify(dataStore).fetchCountriesAndStates(selectedSite.get())
```

wzieba commented on WooCommerce/src/main/res/navigation/nav_graph_orders.xml:
```diff
@@ -647,12 +646,36 @@
+    <action
+        android:id="@+id/action_search_filter_fragment"
+        app:destination="@id/searchFilterFragment" />
+    <fragment
+        android:id="@+id/searchFilterFragment"
+        android:name="com.woocommerce.android.ui.searchfilter.SearchFilterFragment"
+        android:label="searchFilterFragment"
+        tools:layout="@layout/fragment_search_filter">
```
💡 WDYT about adding some slide-in and slide-out animations?

wzieba reviewed (commented):
Hi @RenanLukas 👋  thanks for adding the feature, I've played with it and looks really nice. I left some ideas in comments, none of them is change request though - more like improvements suggestions.

I would ask you to work on one thing though: during testing I've spotted memory leaks. To reproduce it I opened country or state search fragment and returned to address editing fragment few times. Below I'm pasting leak canary report of one of the leaks. Could you please check what's the cause and fix it? Thanks!

```
​
┬───
│ GC Root: Global variable in native code
│
├─ android.graphics.animation.RenderNodeAnimator instance
│    Leaking: UNKNOWN
│    Retaining 437.5 kB in 7847 objects
│    ↓ RenderNodeAnimator.mTarget
│                         ~~~~~~~
├─ android.graphics.RenderNode instance
│    Leaking: UNKNOWN
│    Retaining 435.1 kB in 7769 objects
│    ↓ RenderNode.mAnimationHost
│                 ~~~~~~~~~~~~~~
├─ android.view.ViewAnimationHostBridge instance
│    Leaking: UNKNOWN
│    Retaining 435.0 kB in 7768 objects
│    ↓ ViewAnimationHostBridge.mView
│                              ~~~~~
├─ androidx.constraintlayout.widget.ConstraintLayout instance
│    Leaking: UNKNOWN
│    Retaining 435.0 kB in 7767 objects
│    View not part of a window view hierarchy
│    View.mAttachInfo is null (view detached)
│    View.mWindowAttachCount = 2
│    mContext instance of dagger.hilt.android.internal.managers.ViewComponentManager$FragmentContextWrapper, wrapping
│    activity com.woocommerce.android.ui.main.MainActivity with mDestroyed = false
│    ↓ View.mParent
│           ~~~~~~~
├─ androidx.recyclerview.widget.RecyclerView instance
│    Leaking: UNKNOWN
│    Retaining 374.6 kB in 6244 objects
│    View not part of a window view hierarchy
│    View.mAttachInfo is null (view detached)
│    View.mID = R.id.searchItemsList
│    View.mWindowAttachCount = 1
│    mContext instance of dagger.hilt.android.internal.managers.ViewComponentManager$FragmentContextWrapper, wrapping
│    activity com.woocommerce.android.ui.main.MainActivity with mDestroyed = false
│    ↓ RecyclerView.mAdapter
│                   ~~~~~~~~
├─ com.woocommerce.android.ui.searchfilter.SearchFilterAdapter instance
│    Leaking: UNKNOWN
│    Retaining 148 B in 7 objects
│    ↓ SearchFilterAdapter.onItemSelectedListener
│                          ~~~~~~~~~~~~~~~~~~~~~~
├─ com.woocommerce.android.ui.searchfilter.SearchFilterFragment$setupSearchList$1$1 instance
│    Leaking: UNKNOWN
│    Retaining 16 B in 1 objects
│    Anonymous subclass of kotlin.jvm.internal.Lambda
│    ↓ SearchFilterFragment$setupSearchList$1$1.this$0
│                                               ~~~~~~
╰→ com.woocommerce.android.ui.searchfilter.SearchFilterFragment instance
​     Leaking: YES (ObjectWatcher was watching this because com.woocommerce.android.ui.searchfilter.
​     SearchFilterFragment received Fragment#onDestroy() callback and Fragment#mFragmentManager is null)
​     Retaining 22.0 kB in 880 objects
​     key = 2bd9b70b-982e-4f0a-b9f5-747d6dff1c30
​     watchDurationMillis = 6413
​     retainedDurationMillis = 1404
​     componentContext instance of dagger.hilt.android.internal.managers.ViewComponentManager$FragmentContextWrapper,
​     wrapping activity com.woocommerce.android.ui.main.MainActivity with mDestroyed = false

METADATA

Build.VERSION.SDK_INT: 31
Build.MANUFACTURER: Google
LeakCanary version: 2.7
App process name: com.woocommerce.android.dev
Stats: LruCache[maxSize=3000,hits=11043,misses=210577,hitRate=4%]
RandomAccess[bytes=11356731,reads=210577,travel=119760579125,range=44441203,size=53458354]
Heap dump reason: user request
Analysis duration: 9520 ms
```

RenanLukas committed:
Fix redundant calls to times method on unit tests

RenanLukas committed:
Add enter and exit animations for search filter screen

RenanLukas committed:
Refactor search filter RecyclerView.Adapter to ListAdapter

RenanLukas commented:
Hey @wzieba , thank you very much for the review! I've addressed the improvement suggestions you made.

I'm having some trouble trying to reproduce the memory leak though. I've tried going back and forth multiple times between country/state search and address editing screen. I couldn't get anything on LeakCanary or memory profiler. Could it be that I'm not following the steps to reproduce correctly or maybe it's a flaky issue? 🤔 Either way I'll continue to try to reproduce it.

wzieba commented:
hi @RenanLukas , thank for adding improvements! About the leak, unfortunately I don't have precise steps but here I'm dropping a video that might help in reproduction:

https://user-images.githubusercontent.com/5845095/141843371-17c1ca74-3c23-4a62-970c-515f0baed2e2.mov

RenanLukas committed:
Fix SearchFilterFragment memory leak (search TextWatcher)

RenanLukas commented:
Thank you for recording a video @wzieba. It was really helpful 🎉 

Although I couldn't reproduce it on `LeakCanary`, this time I had more luck with the Memory Profiler. I've found a memory leak related to the search `TextWatcher` (it was never being removed 😬). After handling it, I didn't get any more leaks on Memory Profiler, but since I couldn't reproduce it on `LeakCanary`, could you please try to reproduce it again when you have time?

![leak_memory_profiler_2](https://user-images.githubusercontent.com/14964993/141923265-b94ef7dc-69a4-4bc7-a002-bdf0533e8424.png)

wzieba commented:
Thanks @RenanLukas , good job - I tried hard but couldn't reproduce the memory leak 🎉 .

---

I found one issue though: when we rotate screen, value of `Country` and `State` disappear. WDYT folks @nbradbury @RenanLukas  - should we fix that in scope of this PR or some next one?

https://user-images.githubusercontent.com/5845095/142282244-181406cc-c194-4b98-bdb0-b6014a900aa3.mov

RenanLukas commented:
Good catch @wzieba
A while ago I've tested the current version of this feature (with a Dialog instead of a Fragment) and the same thing happened.   
@nbradbury filed a bug [here](https://github.com/woocommerce/woocommerce-android/issues/5205). If any of you think it's worth to fix the bug in this PR, please let me know and I'll fix it.

https://user-images.githubusercontent.com/14964993/142308940-bbe686e4-f2c8-41a2-ae24-f8b5a39b8fcf.mp4

wzieba commented:
oh sorry, now I see the discussion on Slack. Sorry I missed that. I think it's completely fine to fix this in a separate PR.

Once again, great job on this PR!

RenanLukas commented:
No worries @wzieba 
Thanks for the review 🎉 

@nbradbury @wzieba should I merge it?

nbradbury commented:
@RenanLukas You can merge it to your repo if you like, but we'll need to clone this PR and merge it to our repo since you don't have permission to do that.

RenanLukas commented:
@nbradbury right! I'll just merge to the feature branch on this repo then. Thanks!
